plugins/inputs/opcua/read_client.go

```diff
@@ -158,26 +181,46 @@ func (o *readClient) read() error {
 		if err == nil {
 			// Success, update the node values and exit
 			o.ReadSuccess.Incr(1)
+			o.lastSessionError = false
 			for i, d := range resp.Results {
 				o.UpdateNodeValue(i, d)
 			}
 			return nil
 		}
+
 		o.ReadError.Incr(1)
+		lastErr = err
+
+		isSessionError := errors.Is(err, ua.StatusBadSessionIDInvalid) ||
+			errors.Is(err, ua.StatusBadSessionNotActivated) ||
+			errors.Is(err, ua.StatusBadSecureChannelIDInvalid)
+
+		// Flag session error for next cycle if encountered
+		if isSessionError {
+			o.lastSessionError = true
+		}
```

Comment:
```suggestion
		// Flag session error for next cycle if encountered
		if errors.Is(err, ua.StatusBadSessionIDInvalid) ||
		   errors.Is(err, ua.StatusBadSessionNotActivated) ||
		   errors.Is(err, ua.StatusBadSecureChannelIDInvalid) {
			o.forceReconnect = true
		}
```

Reply:
`isSessionError` is getting used in the switch case as well, that's why I put it in a variable.

Reply:
Looking closer on this, why do you move this out of the `switch` statement? You could set `forceReconnect` in there too.

Reply:
The `if isSessionError` must stay separate from the switch case. We need to set `o.forceReconnect = true` for EVERY session error (including when retries are exhausted), but the `case isSessionError:` only handles retry logic when we haven't exceeded the limit. Moving it into the case would miss setting the flag when `count > o.ReadRetries`.